### PR TITLE
[SD-2025] - Notifications on setting changes

### PR DESCRIPTION
**Pull request by akintolga**
- [x] Notifications on user settings changes
- [x] Notifications on role settings changes
- [x] Notifications on desk user and desk settings changes 
- [x] Notifications on stage settings changes

### CODE REVIEW

File: server/apps/users/services.py [hunk 1]
```diff
@@ -140,6 +151,29 @@ def __handle_status_changed(self, updates, user):
             if can_send_mail:
                 send_user_status_changed_email([user['email']], status)
 
+    def __send_notification(self, updates, user):
+        user_id = user['_id']
+
+        if 'is_enabled' in updates and not updates['is_enabled']:
+            push_notification('LOGOUT:user_disabled', user_id=str(user_id))
+        elif 'is_active' in updates and not updates['is_active']:
+            push_notification('LOGOUT:user_disabled', user_id=str(user_id))
```
Comment (sivakuna-aap) [inline server/apps/users/services.py]: There should be differentiation between Inactive and Disabled notifications

File: server/apps/users/services.py [hunk 2]
```diff
@@ -167,6 +201,8 @@ def on_update(self, updates, original):
         if updates.get('is_enabled', False):
             updates['is_active'] = True
 
+        self.__send_notification(updates, original)
```
Comment (sivakuna-aap) [inline server/apps/users/services.py]: Should be in `on_updated()` instead of `on_update()`

**Comment by marwoodandrew** [comment]
:+1:

File: server/apps/desks.py [hunk 3]
```diff
@@ -118,10 +119,43 @@ def delete(self, lookup):
...
+        desk_id = desk['_id']
+
+        if 'members' in updates:
+            added, removed = self.__compare_members(desk.get('members', {}), updates['members'])
+            if len(removed) > 0:
+                push_notification('desk_membership_revoked',
+                                  updated=1,
+                                  user_ids=[str(item) for item in removed],
+                                  desk_id=str(desk_id))
+            else:
```
Comment (amagdas) [inline server/apps/desks.py]: I think this should happen anyway...

File: server/apps/users/services.py [hunk 4]
```diff
@@ -380,3 +426,25 @@ def remove_old_default(self):
...
         role = self.find_one(req=None, is_default=True)
         return role.get('_id') if role is not None else None
+
+    def on_updated(self, updates, role):
+        self.__send_notification(updates, role)
+
+    def __send_notification(self, updates, role):
+        role_id = role['_id']
+
+        role_users = superdesk.get_resource_service('users').get_users_by_role(role_id)
```
Comment (amagdas) [inline server/apps/users/services.py]: `self.get_users_by_role` should work here

File: server/apps/users/services.py [hunk 5]
```diff
@@ -380,3 +426,25 @@ def remove_old_default(self):
...
+        notified_users = [user['_id'] for user in role_users]
+
+        if 'privileges' in updates:
+            added, removed, modified = compare_preferences(role.get('privileges', {}), updates['privileges'])
+            if len(removed) > 0 or (1, 0) in modified.values():
+                push_notification('role_privileges_revoked', updated=1, role_id=str(role_id))
+            else:
```
Comment (amagdas) [inline server/apps/users/services.py]: We should notify on new privileges being added even if some are removed.

**Comment by amagdas** [comment]
Add some behave tests as well please. :blush:

File: server/apps/users/services.py (hunk 4 above)
Comment (akintolga) [inline server/apps/users/services.py]: Actually no because get_users_by_role is in UsersService, this piece of code is in RolesService.

Comment (amagdas) [inline server/apps/users/services.py]: Ok, sorry, was not visible on the diff.

**Comment by akintolga** [comment]
Added behave tests and fixed the points Adrian mentioned.

**Comment by amagdas** [comment]
:+1:

File: server/features/role.feature [hunk 6]
```diff
@@ -157,3 +175,29 @@ Feature: Role Resource
...
+    Scenario: Change user role
+        Given "roles"
+            """
+            [{"name": "A" }]
+            """
+        Given "users"
+            """
+            [{"username": "foo", "first_name": "Foo", "last_name": "Bar", "email": "foo@bar.org", "is_active": true, "role": "#roles._id#"}]
+            """
+        Given "roles"
```
Comment (sivakuna-aap) [inline server/features/role.feature]: This should be post step. This step will empty the roles collection.

File: server/features/desks.feature [hunk 7]
```diff
@@ -116,7 +116,8 @@ Feature: Desks
         """
 
     @auth
-    Scenario: Cannot delete desk if it is assigned as a default desk to user(s)
+    @notification
+    Scenario: Remove use from desk membership
```
Comment (sivakuna-aap) [inline server/features/desks.feature]: `r` is missing

**Comment by sivakuna-aap** [comment]
:+1: besides my comments

**Pull request merged**